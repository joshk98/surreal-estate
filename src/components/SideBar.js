import React, { useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import qs from "qs";
import { faMagnifyingGlass } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import "../styles/sidebar.css";

const buildQueryString = (search, operation, valueObj) => {
  const currentQueryParams = qs.parse(search, { ignoreQueryPrefix: true });
  const newQueryParams = {
    ...currentQueryParams,
    [operation]: JSON.stringify({
      ...JSON.parse(currentQueryParams[operation] || "{}"),
      ...valueObj,
    }),
  };
  return qs.stringify(newQueryParams, { addQueryPrefix: true, encode: false });
};

const SideBar = ({ search }) => {
  const [query, setQuery] = useState("");
  const navigate = useNavigate();

  const handleSearch = (event) => {
    event.preventDefault();
    const newQueryString = buildQueryString(search, "query", {
      title: { $regex: query },
    });
    navigate(newQueryString);
  };

  return (
    <div className="side-bar__links">
      <form className="search-form" onSubmit={handleSearch}>
        <input
          className="search-input"
          type="text"
          placeholder="Search by title"
          value={query}
          onChange={(e) => setQuery(e.target.value)}
        />
        <button className="search-button" type="submit">
          <FontAwesomeIcon icon={faMagnifyingGlass} />
        </button>
      </form>
      <Link to="/">Show All</Link>
      <Link to={buildQueryString(search, "query", { city: "Manchester" })}>
        Manchester
      </Link>
      <Link to={buildQueryString(search, "query", { city: "Leeds" })}>
        Leeds
      </Link>
      <Link to={buildQueryString(search, "query", { city: "Sheffield" })}>
        Sheffield
      </Link>
      <Link to={buildQueryString(search, "query", { city: "Liverpool" })}>
        Liverpool
      </Link>
      <Link to={buildQueryString(search, "sort", { price: 1 })}>
        Price Ascending
      </Link>
      <Link to={buildQueryString(search, "sort", { price: -1 })}>
        Price Descending
      </Link>
    </div>
  );
};

export default SideBar;
