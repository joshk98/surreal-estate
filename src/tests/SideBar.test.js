import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import SideBar from "../components/SideBar";

describe("SideBar", () => {
  it("renders correctly", () => {
    const { asFragment } = render(
      <Router>
        <SideBar />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
