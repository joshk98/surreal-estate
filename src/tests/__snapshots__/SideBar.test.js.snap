// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar renders correctly 1`] = `
<DocumentFragment>
  <div
    class="side-bar__links"
  >
    <a
      href="/"
    >
      Show All
    </a>
    <a
      href="/?query={\\"city\\":\\"Manchester\\"}"
    >
      Manchester
    </a>
    <a
      href="/?query={\\"city\\":\\"Leeds\\"}"
    >
      Leeds
    </a>
    <a
      href="/?query={\\"city\\":\\"Sheffield\\"}"
    >
      Sheffield
    </a>
    <a
      href="/?query={\\"city\\":\\"Liverpool\\"}"
    >
      Liverpool
    </a>
    <a
      href="/?sort={\\"price\\":1}"
    >
      Price Ascending
    </a>
    <a
      href="/?sort={\\"price\\":-1}"
    >
      Price Descending
    </a>
  </div>
</DocumentFragment>
`;
